streams/src/test/java/org/apache/kafka/streams/integration/IQv2VersionedStoreIntegrationTest.java

```diff
@@ -0,0 +1,198 @@
+    private StreamsBuilder builder;
+    private KafkaStreams kafkaStreams;
+
+    public IQv2VersionedStoreIntegrationTest(){}
+
+    @BeforeClass
+    public static void before()
+        throws InterruptedException, IOException, ExecutionException, TimeoutException {
+        CLUSTER.start();
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        Thread.sleep(10000);
+        final KafkaProducer<Integer, Integer> producer = new KafkaProducer<>(producerProps);
+        producer.send(new ProducerRecord<>(INPUT_TOPIC_NAME, 0, RECORD_TIMESTAMP_LONG, 2, 2)).get();
+    }
+
+    @Before
+    public void beforeTest() throws InterruptedException {
+        builder = new StreamsBuilder();
+        builder.table(INPUT_TOPIC_NAME,
+            Materialized.as(Stores.persistentVersionedKeyValueStore(STORE_NAME, HISTORY_RETENTION)));
+        final Properties configs = new Properties();
+        configs.put(StreamsConfig.APPLICATION_ID_CONFIG, "app");
+        configs.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        configs.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.IntegerSerde.class.getName());
+        configs.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.IntegerSerde.class.getName());
+        kafkaStreams = new KafkaStreams(builder.build(), configs);
+        kafkaStreams.start();
+        Thread.sleep(10000);
+    }
+
+    @After
+    public void afterTest() {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+            kafkaStreams.cleanUp();
+        }
+    }
+
+    @AfterClass
+    public static void after() {
+        CLUSTER.stop();
+    }
+
+    @Test
+    public void verifyStore() {
+        try {
+            final Function<VersionedRecord<Integer>, Integer> valueExtractor =
+                VersionedRecord :: value;
+            final Function<VersionedRecord<Integer>, Long> timestampExtractor =
+                VersionedRecord :: timestamp;
+            shouldHandleVersionedKeyQuery(2, valueExtractor, 2);
+            shouldHandleVersionedKeyQueryWithTS(2, valueExtractor, 2, timestampExtractor, RECORD_TIMESTAMP_LONG);
+            shouldThrowNPEWithNullKey();
+
+        } catch (final AssertionError e) {
+            LOG.error("Failed assertion", e);
+            throw e;
+        }
+    }
+
+    public void shouldHandleVersionedKeyQuery(
+        final Integer key,
+        final Function<VersionedRecord<Integer>, Integer> valueExtactor,
+        final Integer expectedValue) {
+        final VersionedKeyQuery<Integer, Integer> query = VersionedKeyQuery.withKey(key);
```

Comment:
nit: add empty line -- hard to see where the method signature end, and the actually method body starts (or move `) {` to it's own line.